File: components/portfolio-value-go/export_portfolio_value_valuation/valuation.go
// `portfolio:value`, in Go — the same WIT `components/portfolio-value` exports.
//
// This is the binder app's hardest piece of arithmetic (FIFO cost basis, realised
// and unrealised gain, the series a chart is drawn from) written a second time in a
// language nothing else in this repository uses, and dropped into the SAME
// composition. `examples/binder/tests` is the judge: it asserts the money as
// arithmetic — buy 2 @ 10.00, buy 1 @ 40.00, sell 1 @ 30.00 realises 20.00 under
// FIFO and 10.00 under average cost — so it fails for a plausible wrong answer and
// not only for a broken one.
//
// Deliberately a re-derivation, not a translation: the point is that two
// independent implementations agree at the boundary, which is a claim a
// transliteration cannot make.
//
// The rules it has to honour, all of which the specification checks:
//
//   - FIFO. A sale consumes the OLDEST unsold lot, at that lot's own cost. Average
//     cost is easier and answers a different question than the one a collector asks.
//   - Events are SORTED by `at`, never trusted in the order given, so backfilling
//     an old purchase does not change the answer. Stable, so two events in the same
//     second keep their given order — Rust's `sort_by_key` is stable too, and an
//     unstable sort here would disagree with it exactly on the ties that matter.
//   - An unpriced holding is carried at COST and counted, not valued at zero and
//     not dropped. Either of those puts a confident wrong number on the chart.
//   - Money never converts. Two currencies are an error naming both.
package export_portfolio_value_valuation

import (
	"sort"

	witTypes "go.bytecodealliance.org/pkg/wit/types"
	w "wit_component/portfolio_value_valuation"
)

type lot struct {
	quantity  uint32
	unitMinor int64
}

func errValuation(e w.ValueError) witTypes.Result[w.Valuation, w.ValueError] {
	return witTypes.Err[w.Valuation, w.ValueError](e)
}

// ValueAt is what the collection is worth at `at`, over `events` priced by `quotes`.
//
// Events after `at` are ignored, which is also how the history is walked.
func ValueAt(events []w.Event, quotes []w.Quote, at uint64) witTypes.Result[w.Valuation, w.ValueError] {
	if len(events) == 0 {
		// No events means no currency to report a zero IN, so there is no answer.
		return errValuation(w.MakeValueErrorEmpty())
	}

	currency := events[0].Currency
	for _, e := range events {
		if e.Quantity == 0 {
			return errValuation(w.MakeValueErrorZeroQuantity(
				witTypes.Tuple2[string, uint64]{F0: e.ItemId, F1: e.At}))
		}
		if e.Currency != currency {
			return errValuation(w.MakeValueErrorMixedCurrency(
				witTypes.Tuple2[string, string]{F0: currency, F1: e.Currency}))
		}
	}

	sorted := make([]w.Event, len(events))
	copy(sorted, events)
	sort.SliceStable(sorted, func(i, j int) bool { return sorted[i].At < sorted[j].At })

	lots := map[string][]lot{}
	// Insertion order, so the walk below is deterministic. A Go map's range order
	// is randomised on purpose; the sums are order-independent but `unquoted` and
	// any future short-circuit are not, and relying on that is how a test starts
	// passing four times in five.
	var ids []string
	seen := map[string]bool{}
	touch := func(id string) {
		if !seen[id] {
			seen[id] = true
			ids = append(ids, id)
		}
	}

	var realised int64
	for _, e := range sorted {
		if e.At > at {
			continue
		}
		touch(e.ItemId)
		switch e.Kind {
		case w.EventKindAcquired:
			lots[e.ItemId] = append(lots[e.ItemId], lot{e.Quantity, e.UnitMinor})
		case w.EventKindDisposed:
			var held uint32
			for _, l := range lots[e.ItemId] {
				held += l.quantity
			}
			if held < e.Quantity {
				// The event log is wrong. Guessing which event is the lie is a
				// bigger lie than refusing.
				return errValuation(w.MakeValueErrorOversoldAt(
					witTypes.Tuple4[string, uint64, uint32, uint32]{
						F0: e.ItemId, F1: e.At, F2: held, F3: e.Quantity}))
			}
			remaining := e.Quantity
			var cost int64
			dq := lots[e.ItemId]
			for remaining > 0 {
				if dq[0].quantity <= remaining {
					cost += int64(dq[0].quantity) * dq[0].unitMinor
					remaining -= dq[0].quantity
					dq = dq[1:]
				} else {
					cost += int64(remaining) * dq[0].unitMinor
					dq[0].quantity -= remaining
					remaining = 0
				}
			}
			lots[e.ItemId] = dq
			realised += int64(e.Quantity)*e.UnitMinor - cost
		}
	}

	var costBasis, marketValue int64
	var unquoted uint32
	for _, id := range ids {
		var held uint32
		var lotCost int64
		for _, l := range lots[id] {
			held += l.quantity
			lotCost += int64(l.quantity) * l.unitMinor
		}
		if held == 0 {
			continue
		}
		costBasis += lotCost

		// The latest quote at or before `at`, in the portfolio's currency. `>=`
		// rather than `>` so ties resolve to the last one given, which is what
		// Rust's `max_by_key` does.
		var best *w.Quote
		for i := range quotes {
			q := &quotes[i]
			if q.ItemId != id || q.At > at || q.Currency != currency {
				continue
			}
			if best == nil || q.At >= best.At {
				best = q
			}
		}
		if best != nil {
			marketValue += int64(held) * best.UnitMinor
		} else {
			marketValue += lotCost
			unquoted += held
		}
	}

	return witTypes.Ok[w.Valuation, w.ValueError](w.Valuation{
		CostBasisMinor:   costBasis,
		MarketValueMinor: marketValue,
		UnrealisedMinor:  marketValue - costBasis,
		RealisedMinor:    realised,
		Currency:         currency,
		Unquoted:         unquoted,
	})
}

// Series is ValueAt sampled every `step` seconds over `since..=until`, for a chart.
func Series(events []w.Event, quotes []w.Quote, since, until, step uint64) witTypes.Result[[]w.Point, w.ValueError] {
	if step == 0 {
		return witTypes.Err[[]w.Point, w.ValueError](w.MakeValueErrorZeroStep())
	}

	points := []w.Point{}
	t := since
	for {
		r := ValueAt(events, quotes, t)
		if r.IsErr() {
			return witTypes.Err[[]w.Point, w.ValueError](r.Err())
		}
		v := r.Ok()
		points = append(points, w.Point{
			At:               t,
			MarketValueMinor: v.MarketValueMinor,
			CostBasisMinor:   v.CostBasisMinor,
			RealisedMinor:    v.RealisedMinor,
			Unquoted:         v.Unquoted,
		})
		if t >= until {
			break
		}
		next := t + step
		if next < t { // saturating: a step that wraps u64 would loop forever
			next = ^uint64(0)
		}
		if next > until {
			next = until
		}
		t = next
	}
	return witTypes.Ok[[]w.Point, w.ValueError](points)
}

File: components/portfolio-value-cs/Valuation.cs
// `portfolio:value`, in C# — the same WIT `components/portfolio-value` exports.
//
// One of four languages this interface is implemented in. It goes into the SAME
// binder composition and is judged by the SAME unedited e2e; see
// ../portfolio-value-go/README.md for what that is meant to prove.
//
// `result<T, E>` is return-or-throw here: the ok value is returned, and the error
// is thrown as `WitException<ValueError>(e, 0)` — the generated interop catches it
// and lowers it into the error arm. The nesting level is 0 because this result is
// not inside another one.
//
// The rules, all of which the app's e2e checks:
//
//   FIFO. A sale consumes the OLDEST unsold lot, at that lot's own cost. Average
//   cost is easier and answers a different question than the one a collector asks.
//
//   Events are SORTED by `at`, never trusted in the order given, so backfilling an
//   old purchase does not change the answer. STABLE, so two events in the same
//   second keep their given order — `List.Sort` is NOT stable, which is why this
//   uses OrderBy: LINQ's is, and disagreeing with the other three implementations
//   on exactly the ties that matter is the failure mode.
//
//   An unpriced holding is carried at COST and counted, not valued at zero and not
//   dropped. Either of those puts a confident wrong number on the chart.
//
//   Money never converts. Two currencies are an error naming both.

using System.Collections.Generic;
using System.Linq;
using PortfolioValueWorld.wit;

namespace PortfolioValueWorld.wit.Exports.portfolio.value;

public class ValuationExportsImpl : IValuationExports {

    /// One open lot: what is left of a purchase, and what that purchase cost.
    private sealed class Lot {
        public uint Quantity;
        public long UnitMinor;
        public Lot(uint quantity, long unitMinor) { Quantity = quantity; UnitMinor = unitMinor; }
    }

    private readonly record struct Totals(long CostBasis, long MarketValue, long Realised, uint Unquoted);

    private static WitException<IValuationExports.ValueError> Fail(IValuationExports.ValueError e) =>
        new WitException<IValuationExports.ValueError>(e, 0);

    /// The numbers, without the currency — so `Series` can sample without
    /// re-deriving the string every time.
    private static Totals TotalsAt(
        List<IValuationExports.Event> events,
        List<IValuationExports.Quote> quotes,
        ulong at) {

        if (events.Count == 0)
            // No events means no currency to report a zero IN, so there is no answer.
            throw Fail(IValuationExports.ValueError.Empty());

        string currency = events[0].currency;
        foreach (var e in events) {
            if (e.quantity == 0)
                throw Fail(IValuationExports.ValueError.ZeroQuantity((e.itemId, e.at)));
            if (e.currency != currency)
                throw Fail(IValuationExports.ValueError.MixedCurrency((currency, e.currency)));
        }

        // Insertion order for the walk below, so the answer never depends on a
        // dictionary's iteration order.
        var order = new List<string>();
        var lots = new Dictionary<string, List<Lot>>();
        var heads = new Dictionary<string, int>();

        long realised = 0;
        foreach (var e in events.OrderBy(e => e.at)) {   // OrderBy is stable; List.Sort is not
            if (e.at > at) continue;
            if (!lots.ContainsKey(e.itemId)) {
                lots[e.itemId] = new List<Lot>();
                heads[e.itemId] = 0;
                order.Add(e.itemId);
            }
            var dq = lots[e.itemId];

            if (e.kind == IValuationExports.EventKind.ACQUIRED) {
                dq.Add(new Lot(e.quantity, e.unitMinor));
                continue;
            }

            uint held = 0;
            for (int i = heads[e.itemId]; i < dq.Count; i++) held += dq[i].Quantity;
            if (held < e.quantity)
                // The event log is wrong, and guessing which event is the lie is a
                // bigger lie than refusing.
                throw Fail(IValuationExports.ValueError.OversoldAt((e.itemId, e.at, held, e.quantity)));

            uint remaining = e.quantity;
            long cost = 0;
            while (remaining > 0) {
                var front = dq[heads[e.itemId]];
                if (front.Quantity <= remaining) {
                    cost += (long)front.Quantity * front.UnitMinor;
                    remaining -= front.Quantity;
                    heads[e.itemId]++;
                } else {
                    cost += (long)remaining * front.UnitMinor;
                    front.Quantity -= remaining;
                    remaining = 0;
                }
            }
            realised += (long)e.quantity * e.unitMinor - cost;
        }

        long costBasis = 0, marketValue = 0;
        uint unquoted = 0;
        foreach (var itemId in order) {
            var dq = lots[itemId];
            uint held = 0;
            long lotCost = 0;
            for (int i = heads[itemId]; i < dq.Count; i++) {
                held += dq[i].Quantity;
                lotCost += (long)dq[i].Quantity * dq[i].UnitMinor;
            }
            if (held == 0) continue;
            costBasis += lotCost;

            // The latest quote at or before `at`, in this currency. `>=` so ties
            // resolve to the LAST one given, matching the other implementations.
            IValuationExports.Quote? best = null;
            foreach (var q in quotes) {
                if (q.itemId != itemId || q.at > at || q.currency != currency) continue;
                if (best is null || q.at >= best.Value.at) best = q;
            }
            if (best is not null) {
                marketValue += (long)held * best.Value.unitMinor;
            } else {
                // Carried at COST and counted: zero makes the chart lie downward,
                // dropping it lies upward.
                marketValue += lotCost;
                unquoted += held;
            }
        }

        return new Totals(costBasis, marketValue, realised, unquoted);
    }

    public static IValuationExports.Valuation ValueAt(
        List<IValuationExports.Event> events,
        List<IValuationExports.Quote> quotes,
        ulong at) {

        var t = TotalsAt(events, quotes, at);
        return new IValuationExports.Valuation(
            t.CostBasis, t.MarketValue, t.MarketValue - t.CostBasis, t.Realised,
            events[0].currency, t.Unquoted);
    }

    public static List<IValuationExports.Point> Series(
        List<IValuationExports.Event> events,
        List<IValuationExports.Quote> quotes,
        ulong since, ulong until, ulong step) {

        if (step == 0) throw Fail(IValuationExports.ValueError.ZeroStep());

        var points = new List<IValuationExports.Point>();
        ulong t = since;
        for (;;) {
            var v = TotalsAt(events, quotes, t);
            points.Add(new IValuationExports.Point(t, v.MarketValue, v.CostBasis, v.Realised, v.Unquoted));
            if (t >= until) break;
            ulong next = t + step;
            if (next < t) next = ulong.MaxValue;   // saturating: a wrapping step never terminates
            if (next > until) next = until;
            t = next;
        }
        return points;
    }
}
